File: src/lib/schedule.ts
import { loadConfig } from "./config";
import { loadRoutine } from "./routine";
import {
  addCalendarDays,
  isoWeekNumber,
  parseHm,
  pythonWeekday,
  toIsoWithOffset,
  zonedParts,
  zonedWallToDate,
} from "./timezone";
import type {
  EnrichedBlock,
  Routine,
  RoutineBlock,
  RoutinePhase,
  ScheduleSnapshot,
} from "./types";

export const DAY_KEYS = ["mon", "tue", "wed", "thu", "fri", "sat", "sun"] as const;

export function dayKeyForParts(year: number, month: number, day: number): string {
  return DAY_KEYS[pythonWeekday(year, month, day)];
}

function parsePlanStart(value: string): { year: number; month: number; day: number } {
  const [year, month, day] = value.split("-").map(Number);
  return { year, month, day };
}

export function blockWindow(
  year: number,
  month: number,
  day: number,
  start: string,
  end: string,
  timeZone: string,
): { startDt: Date; endDt: Date } {
  const startHm = parseHm(start);
  const endHm = parseHm(end);
  const startDt = zonedWallToDate(
    year,
    month,
    day,
    startHm.hour,
    startHm.minute,
    timeZone,
  );
  let endDt = zonedWallToDate(year, month, day, endHm.hour, endHm.minute, timeZone);
  if (endDt.getTime() <= startDt.getTime()) {
    const next = addCalendarDays(year, month, day, 1);
    endDt = zonedWallToDate(
      next.year,
      next.month,
      next.day,
      endHm.hour,
      endHm.minute,
      timeZone,
    );
  }
  return { startDt, endDt };
}

export function resolveSubject(
  block: RoutineBlock,
  year: number,
  month: number,
  day: number,
): string {
  const subject = block.subject || "none";
  if (subject !== "hld_lld_alt") return subject;
  const evenWeek = isoWeekNumber(year, month, day) % 2 === 0;
  const weekday = pythonWeekday(year, month, day);
  if (weekday === 3) return evenWeek ? "hld" : "lld";
  if (weekday === 4) return evenWeek ? "lld" : "hld";
  return "hld";
}

export function resolveTitle(
  block: RoutineBlock,
  year: number,
  month: number,
  day: number,
): string {
  const subject = resolveSubject(block, year, month, day);
  const title = block.title || "Block";
  if (block.subject === "hld_lld_alt") return subject.toUpperCase();
  return title;
}

export function phaseFor(
  planStart: string,
  year: number,
  month: number,
  routine: Routine,
): RoutinePhase {
  const start = parsePlanStart(planStart);
  const monthsElapsed = (year - start.year) * 12 + (month - start.month);
  const monthIndex = Math.max(1, Math.min(6, monthsElapsed + 1));
  const phases = routine.phases || [];
  for (const phase of phases) {
    if (phase.start_month <= monthIndex && monthIndex <= phase.end_month) {
      return { ...phase, month_index: monthIndex };
    }
  }
  const last = phases[phases.length - 1] || {
    name: "Plan",
    start_month: 1,
    end_month: 6,
    mix: "",
  };
  return { ...last, month_index: monthIndex };
}

export function enrichBlock(
  block: RoutineBlock,
  year: number,
  month: number,
  day: number,
  timeZone: string,
  moment: Date,
): EnrichedBlock {
  const { startDt, endDt } = blockWindow(year, month, day, block.start, block.end, timeZone);
  const inBlock = startDt.getTime() <= moment.getTime() && moment.getTime() < endDt.getTime();
  const remaining = inBlock
    ? Math.max(0, Math.floor((endDt.getTime() - moment.getTime()) / 60000))
    : null;
  const elapsed = inBlock
    ? Math.max(0, Math.floor((moment.getTime() - startDt.getTime()) / 60000))
    : null;
  const total = Math.max(1, Math.floor((endDt.getTime() - startDt.getTime()) / 60000));
  const subject = resolveSubject(block, year, month, day);
  return {
    start: block.start,
    end: block.end,
    start_iso: toIsoWithOffset(startDt, timeZone),
    end_iso: toIsoWithOffset(endDt, timeZone),
    title: resolveTitle(block, year, month, day),
    kind: block.kind || "buffer",
    subject,
    guide: block.guide || "",
    minutes: total,
    remaining_min: remaining,
    elapsed_min: elapsed,
    progress_pct:
      elapsed !== null ? Math.min(100, Math.round((100 * elapsed) / total)) : 0,
  };
}

export function iterBlocksForDay(
  routine: Routine,
  year: number,
  month: number,
  day: number,
  timeZone: string,
): EnrichedBlock[] {
  const key = dayKeyForParts(year, month, day);
  const daySpec = routine.days[key];
  const noon = zonedWallToDate(year, month, day, 12, 0, timeZone);
  return daySpec.blocks.map((block) =>
    enrichBlock(block, year, month, day, timeZone, noon),
  );
}

export function currentAndNext(moment = new Date()): ScheduleSnapshot {
  const cfg = loadConfig();
  const routine = loadRoutine();
  const timeZone = cfg.timezone;
  const todayParts = zonedParts(moment, timeZone);
  const yesterday = addCalendarDays(todayParts.year, todayParts.month, todayParts.day, -1);
  const tomorrow = addCalendarDays(todayParts.year, todayParts.month, todayParts.day, 1);

  let current: EnrichedBlock | null = null;
  const upcoming: EnrichedBlock[] = [];

  for (const day of [yesterday, todayParts, tomorrow]) {
    const key = dayKeyForParts(day.year, day.month, day.day);
    const daySpec = routine.days[key];
    for (const raw of daySpec.blocks) {
      const { startDt, endDt } = blockWindow(
        day.year,
        day.month,
        day.day,
        raw.start,
        raw.end,
        timeZone,
      );
      const enriched = enrichBlock(
        raw,
        day.year,
        day.month,
        day.day,
        timeZone,
        moment,
      );
      enriched.day = key;
      enriched.date = `${day.year}-${String(day.month).padStart(2, "0")}-${String(day.day).padStart(2, "0")}`;
      if (startDt.getTime() <= moment.getTime() && moment.getTime() < endDt.getTime()) {
        current = enriched;
      } else if (startDt.getTime() > moment.getTime()) {
        upcoming.push(enriched);
      }
    }
  }

  upcoming.sort((a, b) => a.start_iso.localeCompare(b.start_iso));
  const dayKey = dayKeyForParts(todayParts.year, todayParts.month, todayParts.day);
  const daySpec = routine.days[dayKey];

  return {
    now: toIsoWithOffset(moment, timeZone),
    timezone: timeZone,
    day_key: dayKey,
    day_label: daySpec.label,
    day_kind: daySpec.kind,
    day_summary: daySpec.summary,
    current,
    next: upcoming[0] ?? null,
    upcoming: upcoming.slice(0, 6),
    today: iterBlocksForDay(
      routine,
      todayParts.year,
      todayParts.month,
      todayParts.day,
      timeZone,
    ),
    phase: phaseFor(cfg.planStart, todayParts.year, todayParts.month, routine),
    non_negotiables: routine.non_negotiables || [],
    dsa_explain_flow: routine.dsa_explain_flow || [],
    weekly_hours: routine.weekly_hours || {},
  };
}

File: tracker/schedule.py
from __future__ import annotations

import json
from datetime import date, datetime, time, timedelta
from typing import Any, Dict, List, Optional
from zoneinfo import ZoneInfo

from tracker.config import AppConfig, load_config

DAY_KEYS = ("mon", "tue", "wed", "thu", "fri", "sat", "sun")
DAY_KEY_BY_WEEKDAY = {index: key for index, key in enumerate(DAY_KEYS)}


def load_routine(path=None) -> Dict[str, Any]:
    cfg = load_config()
    routine_path = path or cfg.routine_path
    with routine_path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def now_in_tz(tz_name: str) -> datetime:
    return datetime.now(ZoneInfo(tz_name))


def parse_hm(value: str) -> time:
    hour_s, minute_s = value.split(":")
    hour = int(hour_s)
    minute = int(minute_s)
    if hour == 24:
        return time(0, minute)
    return time(hour, minute)


def block_window(day: date, start: str, end: str, tz_name: str):
    tz = ZoneInfo(tz_name)
    start_dt = datetime.combine(day, parse_hm(start), tzinfo=tz)
    end_dt = datetime.combine(day, parse_hm(end), tzinfo=tz)
    if end_dt <= start_dt:
        end_dt += timedelta(days=1)
    return start_dt, end_dt


def day_key_for(moment: datetime) -> str:
    return DAY_KEY_BY_WEEKDAY[moment.weekday()]


def iso_week_number(moment: datetime) -> int:
    return int(moment.isocalendar()[1])


def resolve_subject(block: Dict[str, Any], moment: datetime) -> str:
    subject = block.get("subject") or "none"
    if subject != "hld_lld_alt":
        return subject
    # Even ISO weeks: Thursday = HLD, Friday = LLD. Odd weeks: the reverse.
    even_week = iso_week_number(moment) % 2 == 0
    weekday = moment.weekday()
    if weekday == 3:  # Thursday
        return "hld" if even_week else "lld"
    if weekday == 4:  # Friday
        return "lld" if even_week else "hld"
    return "hld"


def resolve_title(block: Dict[str, Any], moment: datetime) -> str:
    subject = resolve_subject(block, moment)
    title = block.get("title") or "Block"
    if block.get("subject") == "hld_lld_alt":
        return subject.upper()
    return title


def phase_for(plan_start: date, today: date, routine: Dict[str, Any]) -> Dict[str, Any]:
    months_elapsed = (today.year - plan_start.year) * 12 + (today.month - plan_start.month)
    month_index = max(1, min(6, months_elapsed + 1))
    for phase in routine.get("phases") or []:
        if phase["start_month"] <= month_index <= phase["end_month"]:
            return {**phase, "month_index": month_index}
    last = (routine.get("phases") or [{}])[-1]
    return {**last, "month_index": month_index}


def enrich_block(block: Dict[str, Any], day: date, tz_name: str, moment: datetime) -> Dict[str, Any]:
    start_dt, end_dt = block_window(day, block["start"], block["end"], tz_name)
    remaining = max(0, int((end_dt - moment).total_seconds() // 60)) if start_dt <= moment < end_dt else None
    elapsed = max(0, int((moment - start_dt).total_seconds() // 60)) if start_dt <= moment < end_dt else None
    total = max(1, int((end_dt - start_dt).total_seconds() // 60))
    subject = resolve_subject(block, datetime.combine(day, parse_hm(block["start"]), tzinfo=ZoneInfo(tz_name)))
    return {
        "start": block["start"],
        "end": block["end"],
        "start_iso": start_dt.isoformat(),
        "end_iso": end_dt.isoformat(),
        "title": resolve_title(block, datetime.combine(day, parse_hm(block["start"]), tzinfo=ZoneInfo(tz_name))),
        "kind": block.get("kind") or "buffer",
        "subject": subject,
        "guide": block.get("guide") or "",
        "minutes": total,
        "remaining_min": remaining,
        "elapsed_min": elapsed,
        "progress_pct": int(min(100, round(100 * (elapsed or 0) / total))) if elapsed is not None else 0,
    }


def iter_blocks_for_day(routine: Dict[str, Any], day: date, tz_name: str) -> List[Dict[str, Any]]:
    key = DAY_KEY_BY_WEEKDAY[day.weekday()]
    day_spec = routine["days"][key]
    anchor = datetime.combine(day, time(12, 0), tzinfo=ZoneInfo(tz_name))
    return [enrich_block(block, day, tz_name, anchor) for block in day_spec["blocks"]]


def current_and_next(cfg: AppConfig, moment: Optional[datetime] = None) -> Dict[str, Any]:
    routine = load_routine(cfg.routine_path)
    moment = moment or now_in_tz(cfg.timezone)
    tz_name = cfg.timezone
    today = moment.date()
    yesterday = today - timedelta(days=1)

    current = None
    upcoming: List[Dict[str, Any]] = []

    for day in (yesterday, today, today + timedelta(days=1)):
        key = DAY_KEY_BY_WEEKDAY[day.weekday()]
        day_spec = routine["days"][key]
        for raw in day_spec["blocks"]:
            start_dt, end_dt = block_window(day, raw["start"], raw["end"], tz_name)
            enriched = enrich_block(raw, day, tz_name, moment)
            enriched["day"] = key
            enriched["date"] = day.isoformat()
            if start_dt <= moment < end_dt:
                current = enriched
            elif start_dt > moment:
                upcoming.append(enriched)

    upcoming.sort(key=lambda item: item["start_iso"])
    day_spec = routine["days"][day_key_for(moment)]
    return {
        "now": moment.isoformat(),
        "timezone": tz_name,
        "day_key": day_key_for(moment),
        "day_label": day_spec["label"],
        "day_kind": day_spec["kind"],
        "day_summary": day_spec["summary"],
        "current": current,
        "next": upcoming[0] if upcoming else None,
        "upcoming": upcoming[:6],
        "today": iter_blocks_for_day(routine, today, tz_name),
        "phase": phase_for(cfg.plan_start, today, routine),
        "non_negotiables": routine.get("non_negotiables") or [],
        "dsa_explain_flow": routine.get("dsa_explain_flow") or [],
        "weekly_hours": routine.get("weekly_hours") or {},
    }
